### PR TITLE
Integration between Auth0 Universal Login and Willow

### PR DESCRIPTION
**Pull request by gcotelli**

### CODE REVIEW

File: source/Willow-Auth0-Tests/Auth0UniversalLoginAuthenticationFilterTest.class.st [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+
+{ #category : #running }
+Auth0UniversalLoginAuthenticationFilterTest >> setUp [
+
+	| count |
+
+	super setUp.
+	count := 0.
+	filter := Auth0UniversalLoginAuthenticationFilter
+		withDomain: self domain
```
Comment (mtabacman) [inline source/Willow-Auth0-Tests/Auth0UniversalLoginAuthenticationFilterTest.class.st]: `withDomain:` could be changed to `at:`
`clientId` could be changed to `for:`

File: source/Willow-Auth0-Tests/WillowAuth0DemoTest.class.st [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	^ Array with: registry with: session
+]
+
+{ #category : #running }
+WillowAuth0DemoTest >> setUp [
+
+	super setUp.
+	OSEnvironment current at: 'WA0_CLIENT_SECRET' ifAbsentPut: [ 'secret' ].
+	OSEnvironment current at: 'WA0_DOMAIN' ifAbsentPut: [ 'example.auth0.com' ].
+	OSEnvironment current at: 'WA0_CLIENT_ID' ifAbsentPut: [ 'id' ]
```
Comment (mtabacman) [inline source/Willow-Auth0-Tests/WillowAuth0DemoTest.class.st:38]: There's no `tearDown` to recover the status of these variables after the test is finished.

File: docs/Demo.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+To try the demo you will need an [Auth0](https://auth0.com) account and to configure a new `Regular Web Application`, with the following configuration:
+- http://localhost:3000/willow-auth0-demo in **Allowed Callback URLs**
+- http://localhost:3000/willow-auth0-demo in **Allowed Logout URLs**
+- `HS256` as **JsonWebToken Signature Algorithm** in the Advanced Settings section
+- `Implicit` checked in **Grant Types**
+- Copy the provided `.env` file to the working directory of your image and complete the environment variables with valid settings for your application (you can copy them from the Settings tab in Auth0)
+
+In an image with the `Development` group loaded:
+- Run your image with: `env $(cat .env | xargs) ./pharo-ui`
```
Comment (fortizpenaloza) [inline docs/Demo.md]:
```suggestion
- Run your image with: `env $(cat .env | xargs) ./pharo-ui Pharo.image`
```

Comment (fortizpenaloza) [reply docs/Demo.md]: It'll work only in Linux if you don't specify the image

File: source/Willow-Auth0/WillowApplicationAuthenticatedByAuth0UniversalLogin.class.st [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	"You can copy it from Auth0 Dashboard -> Applications -> Your app -> Client ID"
+
+	^ self subclassResponsibility
+]
+
+{ #category : #private }
+WillowApplicationAuthenticatedByAuth0UniversalLogin class >> jsonWebTokenSignatureAlgorithm [
+
+	"You can copy it from Auth0 Dashboard -> Applications -> Your app -> Advanced Settings -> OAuth ->  JsonWebToken Signature Algorithm
+	and use JsonWebAlgorithm named: 'algorithm name' with the copied value"
```
Comment (mtabacman) [inline source/Willow-Auth0/WillowApplicationAuthenticatedByAuth0UniversalLogin.class.st]: maybe separate the code suggestion to a separate line to make it clearer

**Review by mtabacman — APPROVED** [review]

**Review by fortizpenaloza — DISMISSED** [review]

File: source/Willow-Auth0-Tests/WillowAuth0DemoTest.class.st (hunk 2 above)
Comment (gcotelli) [reply source/Willow-Auth0-Tests/WillowAuth0DemoTest.class.st:38]: Yes, It doesn't matter. There were undefined in the first place.

File: source/Willow-Auth0/WillowApplicationAuthenticatedByAuth0UniversalLogin.class.st [hunk 5]
```diff
@@ -0,0 +1,128 @@
+"
+I'm a Willow application that integrates its authentication with Auth0 Universal Login.
+I will provide a request filter performing the needed authentication redirections, and facilities to obtain the user profile and logout.
+
+I'm an abstract class, my subclasses must provide the following information: Auth0 domain, the client id, the redirection URI and a nonce generation strategy.
+"
+Class {
+	#name : #WillowApplicationAuthenticatedByAuth0UniversalLogin,
```
Comment (jvanecek) [inline source/Willow-Auth0/WillowApplicationAuthenticatedByAuth0UniversalLogin.class.st:8]: Looks too long, maybe AuthenticatedByAuth0UniversalLoginApplication ?

**Review by jvanecek — COMMENTED** [review]
Not very fond of the idea of inheritance for this feature. If I didn't misunderstood the usage, the existing willow applications we have will need to change its superclass to `WillowApplicationAuthenticatedByAuth0UniversalLogin`, just so it adds the filter `newAuthenticationFilter` to the seaside app, right? 
Wouldn't it be better more of a composition scheme? maybe with a builder for the `Auth0UniversalLoginAuthenticationFilter` and so the only change in the existing willow applications is in its #registerAsApplicationUsing: adding the new filter by themselves.

**Comment by gcotelli** [comment]
> Not very fond of the idea of inheritance for this feature. If I didn't misunderstood the usage, the existing willow applications we have will need to change its superclass to `WillowApplicationAuthenticatedByAuth0UniversalLogin`, just so it adds the filter `newAuthenticationFilter` to the seaside app, right?
> Wouldn't it be better more of a composition scheme? maybe with a builder for the `Auth0UniversalLoginAuthenticationFilter` and so the only change in the existing willow applications is in its #registerAsApplicationUsing: adding the new filter by themselves.

The main problem is that you don't have control over the application instantiation. It's made by Seaside in a complex way applying the configuration. Registering the filter is one part, you also need the changes in `initialRequest:` , and you need the other configuration to be attached to the wrapping app so you can't delegate the register* methods. If we have better control on how to instantiate the applications, for sure I would use composition here.

**Comment by serpi90** [comment]
> > Not very fond of the idea of inheritance for this feature. If I didn't misunderstood the usage, the existing willow applications we have will need to change its superclass to `WillowApplicationAuthenticatedByAuth0UniversalLogin`, just so it adds the filter `newAuthenticationFilter` to the seaside app, right?
> > Wouldn't it be better more of a composition scheme? maybe with a builder for the `Auth0UniversalLoginAuthenticationFilter` and so the only change in the existing willow applications is in its #registerAsApplicationUsing: adding the new filter by themselves.
> 
> The main problem is that you don't have control over the application instantiation. It's made by Seaside in a complex way applying the configuration. Registering the filter is one part, you also need the changes in `initialRequest:` , and you need the other configuration to be attached to the wrapping app so you can't delegate the register* methods. If we have better control on how to instantiate the applications, for sure I would use composition here.
I think inheritance fits good for a first approach, but the class name is long as said above.

I would try to change for a composition approach when we have another use case which allows combining UniversalLogin and the new use case.

Regarding the change to existing applications, they will need to change anyways, you will need to discard current login views, and change the superclass or add the composition code.

**Review by serpi90 — APPROVED** [review]

File: source/Willow-Auth0/WillowApplicationAuthenticatedByAuth0UniversalLogin.class.st (hunk 5 above)
Comment (serpi90) [reply source/Willow-Auth0/WillowApplicationAuthenticatedByAuth0UniversalLogin.class.st:8]: `WillowAuth0Application` ?

**Comment by jvanecek** [comment]
> > Not very fond of the idea of inheritance for this feature. If I didn't misunderstood the usage, the existing willow applications we have will need to change its superclass to `WillowApplicationAuthenticatedByAuth0UniversalLogin`, just so it adds the filter `newAuthenticationFilter` to the seaside app, right?
> > Wouldn't it be better more of a composition scheme? maybe with a builder for the `Auth0UniversalLoginAuthenticationFilter` and so the only change in the existing willow applications is in its #registerAsApplicationUsing: adding the new filter by themselves.
> 
> The main problem is that you don't have control over the application instantiation. It's made by Seaside in a complex way applying the configuration. Registering the filter is one part, you also need the changes in `initialRequest:` , and you need the other configuration to be attached to the wrapping app so you can't delegate the register* methods. If we have better control on how to instantiate the applications, for sure I would use composition here.

Yes, I saw the initialRequest:, and I was thinking about adding a flag to WillowApplication that could be setted with for example a message #ensureInitialRequestIsPOST, kind of generic, and the "Auth0AddOn" would add the filter and set this flag. But we can discuss it when have more examples.

**Pull request merged**